✅ Expressions tests and Layers View Improvements

aldebaran30701 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix`, `hotfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change
This should be merged in AFTER #1002 
This PR adds:
- ExpressionParserNGTest.java which is responsible for testing many of the core attributes within various expression strings. It purely tests that the parser returns the correct order and expression types to the caller. I've added checks for invalid queries also. So far, every operator has been tested in the cases, It was required that a flag be set within the test cases to ensure no pop-up was activated during test execution. It should run as intended normally.
- I've fixed a bug with the comparison of expressions which is used within the test cases (AssertEquals(...) and so on). It was to do with equals and hashcode implementations not existing.
- Updated the layers-view.html documentation page with extra information and better formatting.
- Relocated the Expressions specific documentation to the Graph framework module - adding a link from Layers documentation to help assist the user to find the correct help documentation between modules.
- Added tooltips to the buttons of layers view to show the hotkeys for that action.
- Added word operators for most common operators. (equals, and, xor, subtract etc...)
- Implemented an initial parse check on the entry of text into the layers view. This will check that there are no illlegal characters in the query, and that it contains pairs of quotes. Validation will be expensive if the check was to run on the whole graph to see if any elements would be returned.

A red box appears around the text field when it is not valid.
I wanted to include some sort of message that wouldn't interrupt the flow of the Layers View, so I included a context menu to appear. This goes away with a click away.
In normal use the user should really only be concerned about one box, and having multiple context menus appear on different inputs shouldn't be an issue. 
I'm okay to get rid of the context menu if it seems intrusive - keen to hear peoples' thoughts on this.

![image](https://user-images.githubusercontent.com/58677759/122871857-c59a0200-d36e-11eb-94ce-cb9f9d2b07d4.png)

<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
N/A
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Reduces bugs, increases documentation and testing.
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
Reduces bugs, increases documentation and testing.
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
None realized
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
Opening graphs, toggling layers with and without manual entries and query entries. 
Purposefully adding incorrect terms to the UI. Such as adding an extra quote, missing an end quote.
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
Once approved, should solve #970 apart from deletion of a layer. 

<!-- Link any applicable issues here -->

arcturus2 commented:
@aldebaran30701 I'm yet to test this but loving the changes so far. Something I wanted to ask now that you are adding UX things, can the boxes in the expressions section be increased in height by say x5 times? I find that as soon as the scroll bar appears it is really hard to see or make changes.

aldebaran30701 commented:
@arcturus2 I agree - I haven't thought of a good way to counteract this yet.
Here is with 100px height. I tried without wrapping text and it was just a line of text with a scroll bar still. This is with text wrap on.
![image](https://user-images.githubusercontent.com/58677759/123010247-60d9b880-d3fd-11eb-897e-5d1fe7962389.png)

I'd say it's harder to read - I tried using the panel which is similar to the timeline view - and it appears to display better.
![image](https://user-images.githubusercontent.com/58677759/123010646-27ee1380-d3fe-11eb-98a6-1e72da94b558.png)

Here is multiple lines of text - vertical scroll bar appears when more area is needed.
![image](https://user-images.githubusercontent.com/58677759/123010804-769bad80-d3fe-11eb-917b-457275bcb3ba.png)

An issue I see with the timeline-esque view is that the lower buttons are hidden and many of the other layers are not visible. I can 'float' the other buttons but the layers not being visible is another story

aldebaran30701 commented:
I did a quick mock-up of a UI if we were looking to keep it on the vertical bar - Needs some improvement but it gives more space.
![image](https://user-images.githubusercontent.com/58677759/123015835-6dafd980-d408-11eb-9065-416aea427146.png)

arcturus2 commented:
Nice work @aldebaran30701 with the designs.

With regards to putting the layers on the bottom section like the timeline/table view, if you want the buttons to stay they you could look at how the Data Access View does it to "anchor" the buttons to the top (or you could do bottom). It's using an `AnchorPane`.

Regarding the second option with the Mock Up UI I like that too and I think that could work. One suggestion there is that you could introduce a scroll bar for the list of 15 layers visible there to maybe 5 and then use the rest of the screen real-estate to have a section for the "Vertex Query" going horizontally using the full width and have the "Transaction Query" block underneath?

Keep the design you have but if you can put this as another variation I can try to find some people to do a UX study on showing them different mock-ups and see what they think?

Also this can be done in a separate branch so it does not block this stuff getting out. That said I found an issue which I will post in a separate comment.

arcturus2 commented:
@aldebaran30701 so I actually merged both PRs into my own local branch to test these 2 together and I got an error. Can you confirm>

Reproducer is I created 4 nodes in edit mode and in the layer 1 Vertex query i put in `Type == 'Unknown'`.

I got the following exception:

```
au.gov.asd.tac.constellation.graph.value.expression.ExpressionException: Unable to find implementation of operator: EQUALS (au.gov.asd.tac.constellation.graph.attribute.AbstractObjectAttributeDescription$$Lambda$1235/0x0000000800ad4840, au.gov.asd.tac.constellation.graph.value.expression.ExpressionCompiler$$Lambda$1244/0x0000000800add840)
	at au.gov.asd.tac.constellation.graph.value.expression.ExpressionCompiler.compileSequenceExpression(ExpressionCompiler.java:113)
	at au.gov.asd.tac.constellation.views.layers.query.Query.compile(Query.java:89)
	at au.gov.asd.tac.constellation.views.layers.query.BitMaskQuery.update(BitMaskQuery.java:91)
	at au.gov.asd.tac.constellation.views.layers.query.BitMaskQueryCollection.update(BitMaskQueryCollection.java:103)
	at au.gov.asd.tac.constellation.views.layers.query.BitMaskQueryCollection.updateBitMasks(BitMaskQueryCollection.java:121)
	at au.gov.asd.tac.constellation.views.layers.LayersViewController$UpdateQueryPlugin.edit(LayersViewController.java:376)
	at au.gov.asd.tac.constellation.plugins.templates.SimpleEditPlugin.describedEdit(SimpleEditPlugin.java:207)
[catch] at au.gov.asd.tac.constellation.plugins.templates.SimpleEditPlugin.run(SimpleEditPlugin.java:115)
Caused: java.lang.RuntimeException
	at au.gov.asd.tac.constellation.plugins.templates.SimpleEditPlugin.run(SimpleEditPlugin.java:135)
[catch] at au.gov.asd.tac.constellation.graph.node.plugins.DefaultPluginEnvironment.lambda$executePluginLater$0(DefaultPluginEnvironment.java:129)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

arcturus2 reviewed (changes requested):
Please see my last comment of an issue I got.

arcturus2 commented:
I think we can have a ticket to discuss the UI and come up with proposals if you like but here is a taste of what I was thinking:

![image](https://user-images.githubusercontent.com/39325530/123264364-4c9cd500-d53d-11eb-85f6-af59de300969.png)

I'm just using the free mockplus.com site and here is a share link https://app.mockplus.com/app/share-b3f2b17418d4af22c5daf78e28663d18share-EyQFB_dpBbD/rp

aldebaran30701 commented:
I had the same thoughts with the way you did that mockup.
Regarding the error you found - I actually based this branch off the branch that fixed the type issue. Maybe merging it back into this branch caused an issue?
Come to think of it - this PR could be the only one merged in, and the types PR #1111 just closed. I've got all changes in this current PR.

I managed to successfully query all default types in Constellation.
I've also removed the context menu, and added a label down the bottom which toggles it's visibility with a static query error message.
(don't mind the small text - I needed more real estate)
![image](https://user-images.githubusercontent.com/58677759/123356675-4e9b7e00-d5a7-11eb-955f-dafe2c63c32d.png)

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/LayersViewPane.java:
```diff
@@ -283,12 +332,26 @@ private synchronized void syncLayers() {
                     visible = ((CheckBox) item).isSelected();
                     break;
-                case 2:
+                case 2: {
```
omit the braces from the case blocks as they're not necessary

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/docs/layers-view.html:
```diff
@@ -7,6 +7,16 @@
     </head>
     <body>
         <h1>Layers View</h1>
+        <p>
+            The Layers View holds a collection of Layers. Each Layer can represent a static set of elements, or a dynamically 
+            calculated set of elements which match a query criteria. The query structure is based on the expressions framework.
+            <br/>
+            Each layer has a visibility checkbox which determines if that layer is toggled. All elements are on the base layer 0.
+            When one or more check boxes are selected, only the elements that are members of those layers are visibly displayed. 
+            All other elements will have their <strong>visibility</strong> attribute value toggled to 0.0 (invisible).
+            <br/>
+            Arrangements on graphs with a visual schema will be ran on the base layer 0. 
+        </p>
```
Are you able to move this paragraph below the table?

aldebaran30701 replied:
How does this look? Should it be below the image also?
![image](https://user-images.githubusercontent.com/58677759/123561886-38313480-d7ea-11eb-9e97-e5f57c5d1d54.png)

antares1470 replied:
I think I'm ok with the paragraph being above the image as it aligns with what has been done in some of the other pages

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/docs/layers-view.html:
```diff
@@ -7,6 +7,16 @@
+            <br/>
+            Each layer has a visibility checkbox which determines if that layer is toggled. All elements are on the base layer 0.
+            When one or more check boxes are selected, only the elements that are members of those layers are visibly displayed. 
+            All other elements will have their <strong>visibility</strong> attribute value toggled to 0.0 (invisible).
```
With the updates I've made to the help pages, I've actually moved away from using the strong tag just to avoid clashes with title names (which are also bold). I've moved to using double-quotes for now so if you're able to do that as well, that would be great

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/docs/layers-view.html:
```diff
@@ -11,16 +11,16 @@ <h1>Layers View</h1>
             <tr>
                 <th scope="col">
-                    <strong>Constellation Action</strong>
+                    "Constellation Action"
```
I think the double quotes can be removed in the table headers (my bad for not explaining for not explaining what I did with strong tags well enough)

antares1470 commented on CoreLayersView/src/au/gov/asd/tac/constellation/views/layers/docs/layers-view.html:
```diff
@@ -61,180 +61,73 @@ <h1>Layers View</h1>
-        <p>e.g. Label == 'Vertex #0&lt;Unknown&gt;' || Label == 'Vertex #1&lt;Unknown&gt;' shows elements that satisfy either constraint.</p>
-        <p>e.g. Label == 'Vertex #0&lt;Unknown&gt;' &AMP; nradius == '1.5' shows elements that satisfy both constraints.</p>
-    </body>
+        <p>"NOTE: Layers View does not allow for the use of the assignment operator = or keyword assign."</p>
```
Can also remove the double quotes around the Note here (once again my bad)

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/docs/expressions.html:
```diff
@@ -0,0 +1,225 @@
+        <h1>The Constellation Expressions Framework</h1>
+        <p>
+            The expressions framework is a structure which can be used to query or assign the Vertex and Transaction attributes within a graph.
+            The framework will work on all attributes which <strong>do not</strong> include a restricted character within it's AttributeName. An expression can 
+            return <strong>0..*</strong> (none or many) matches.
```
- remove strong tags
- replace "return 0..* (none or many)" with simply "returns zero or more"
- surround AttributeName in double-quotes

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/docs/expressions.html:
```diff
@@ -0,0 +1,225 @@
+        </p>
+        
+        <h2>Forming a Query Expression</h2>
+        <p> A valid query for a vertex or transaction takes the form of <strong>[AttributeName] [Operator] [value]</strong>. </p>
+        <p>The <strong>AttributeName</strong> used will not include any restricted characters, and will be case sensitive.</p>
+        <p>The <strong>AttributeName</strong> will be represented as plain, unquoted text.</p>
+        <p>The <strong>Operator</strong> must be separated by atleast one whitespace character.</p>
+        <p>The <strong>Operator</strong> must be valid and from the list specified below.</p>
+        <p>The <strong>Value</strong> must be quoted using single('value') or double quotes("value").</p>
```
- replace all the strong tags here with double-quotes
- for all bar the first paragraph, I reckon convert the paragraphs to an unordered list

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/docs/expressions.html:
```diff
@@ -0,0 +1,225 @@
+        <p>The <strong>Value</strong> must be quoted using single('value') or double quotes("value").</p>
+        <br/>
+        <p>
+            How to form an expression to find a vertex with the <strong>Label</strong> attribute matching <strong>"Vertex #1&lt;Unknown&gt;"</strong>.
```
bar the table below, I think you can swap all the remaining strong tags with double quotes

aldebaran30701 replied:
So leave the strong tags within the table headers?

antares1470 replied:
I've discovered that table headers are also bold so I don't think they can be removed (as they're not actually adding anything extra)

antares1470 commented on CoreGraphFramework/src/au/gov/asd/tac/constellation/graph/docs/expressions.html:
```diff
@@ -9,8 +9,8 @@
             The expressions framework is a structure which can be used to query or assign the Vertex and Transaction attributes within a graph.
-            The framework will work on all attributes which <strong>do not</strong> include a restricted character within it's AttributeName. An expression can 
-            return <strong>0..*</strong> (none or many) matches.
+            The framework will work on all attributes which do not include a restricted character within it's "AttributeName". An expression can 
```
its*

antares1470 reviewed (changes requested):
Missed that typo in my last review. The only thing I can add is I think it would be good to be able to specify the type without having to specify the full hierarchy. I don't think it's intuitive to have to specify "Type == 'Online Identifier.Email'" rather than "Type == 'Email'" (especially since you don't need to specify the hierarchy in other places in the application)

aldebaran30701 commented:
>The only thing I can add is I think it would be good to be able to specify the type without having to specify the full hierarchy. I don't think it's intuitive to have to specify "Type == 'Online Identifier.Email'" rather than "Type == 'Email'" (especially since you don't need to specify the hierarchy in other places in the application)


@antares1470 The way the expressions framework is extended, it only allows behaviour changes on the operator parsed in, and the type of the attribute (integer, long, double, object etc.) So for the Type attribute, it is an AbstractObjectAttributeDescription. This just uses the implemented toString() method to compare the values of the Type. 

One way forward is to override the toString() method within SchemaElementType, or more specifically the SchemaVertexType & SchemaTransactionType. This, depending on it's usage will cause behavioral changes throughout the whole application.
The file output doesn't seem to use the hierarchy toString(). Which means maybe a change won't break old files.
The other thing I noticed was that the iconmanager uses the toString() of the type to find it's icon. Currently only the unknown Icon seems to have an icon registered with it's type name.

aldebaran30701 commented:
@antares1470 I've found a way to fix this, there is a method which creates the readable object for the expressions framework. This returns the object which usually just calls the default toString. Implementing it within the VertexTypeAttributeDescription allowed me to cast it to a schemaVertextType and then call getname(). commit coming soon

arcturus2 commented:
@aldebaran30701 besides the feedback on the Type string this is working really well. The issues I found earlier are sorted. This PR has come a long way and a lot of effort has gone into this. I can't wait to take this off experimental probably after a GUI update we've been chatting about here.

Also top work on the unit tests here - a stand out PR. Love your work!

antares1470 commented:
Other than resolving merge conflicts, this looks good to go

aldebaran30701 commented:
@antares1470 Should be no conflicts now - Looks good to merge
